src/cpu_treeshap_model.h

```diff
@@ -0,0 +1,358 @@
+  if (fvalue < 0 || std::fabs(fvalue) > max_representable_int) {
+    is_matching_category = false;
+  } else {
+    const auto category_value = static_cast<std::uint32_t>(fvalue);
+    is_matching_category = (
+        std::find(matching_categories.begin(), matching_categories.end(), category_value)
+        != matching_categories.end());
+  }
+  if (categories_list_right_child) {
+    return !is_matching_category;
+  } else {
+    return is_matching_category;
+  }
+}
+
+template <typename tl_threshold_t, typename tl_output_t>
+bool
+decision(
+    const treelite::Tree<tl_threshold_t, tl_output_t>& tree, const float fvalue,
+    int node)
+{
+  // Missing
+  if(std::isnan(fvalue)){
+    return tree.DefaultChild(node) == tree.LeftChild(node);
+  }
+
+  if (tree.SplitType(node) == treelite::SplitFeatureType::kCategorical) {
+    return decision_categorical(
+        fvalue, tree.MatchingCategories(node),
+        tree.CategoriesListRightChild(node));
+  } else {
+    return decision_non_categorical(fvalue, tree.Threshold(node), tree.ComparisonOp(node));
+  }
+}
+
+
+template <typename tl_threshold_t, typename tl_output_t>
+void inference(const TreeMetaInfo<tl_threshold_t, tl_output_t> &tree_info,
+               const float *x,
+               uint8_t *activation,
+               float *value,
+               float *C,
+               float *E,
+               int node = 0,
+               int feature = -1,
+               int depth = 0)
+{
+    const auto& tree = tree_info.tree;
+    float s = 0.;
+    int parent = tree_info.parents[node];
+    if (parent >= 0)
+    {
+        activation[node] = activation[node] & activation[parent];
+        if (activation[parent]) {
+          s = 1 / tree_info.weights[parent];
+        }
+    }
+
+    float *current_e = E + depth * tree_info.max_depth;
+    float *child_e = E + (depth + 1) * tree_info.max_depth;
+    float *current_c = C + depth * tree_info.max_depth;
+    float q = 0.;
+    if (feature >= 0) {
+      if (activation[node]) {
+        q = 1 / tree_info.weights[node];
+      }
+
+      float *prev_c = C + (depth - 1) * tree_info.max_depth;
+      for (int i = 0; i < tree_info.max_depth; i++) {
+        current_c[i] = prev_c[i] * (kBase[i] + q);
+      }
+
+      if (parent >= 0) {
+        for (int i = 0; i < tree_info.max_depth; i++) {
+          current_c[i] = current_c[i] / (kBase[i] + s);
+        }
+      }
+    }
+    int offset_degree = 0;
+    int left = tree.LeftChild(node);
+    int right = tree.RightChild(node);
+    if (!tree.IsLeaf(node)) {
+      activation[left] = decision(tree, x[tree.SplitIndex(node)], node);
+      activation[right] = !activation[left];
+      inference(
+          tree_info, x, activation, value, C, E, left, tree.SplitIndex(node),
+          depth + 1);
+      offset_degree =
+          tree_info.edge_heights[node] - tree_info.edge_heights[left];
+      times_broadcast(kOffset[offset_degree], child_e, tree_info.max_depth);
```

Comment:
nice, by normalizing the polynomials with offset_degree, you don't need to pick which child to traverse first.